python/python/lance/sampler.py

```diff
@@ -105,12 +107,63 @@ def _efficient_sample(
         del tbl
 
 
+def _filtered_efficient_sample(
+    dataset: lance.LanceDataset,
+    n: int,
+    columns: Optional[Union[List[str], Dict[str, str]]],
+    batch_size: int,
+    target_takes: int,
+    filter: str,
+) -> Generator[pa.RecordBatch, None, None]:
+    total_records = len(dataset)
+    shard_size = math.ceil(n / target_takes)
+    num_shards = math.ceil(total_records / shard_size)
+
+    shards = list(range(num_shards))
+    random.shuffle(shards)
+
+    tables = []
+    remaining_rows = n
+    remaining_in_batch = min(batch_size, n)
+    for shard in shards:
+        start = shard * shard_size
+        end = min(start + shard_size, total_records)
+        table = dataset.to_table(
+            columns=columns,
+            offset=start,
+            limit=(end - start),
+            batch_size=shard_size,
+        )
+        if len(columns) == 1 and filter.lower() == f"{columns[0]} is not null":
+            table = pc.drop_null(table)
+        elif filter is not None:
+            raise Exception(f"Can't yet run filter <{filter}> in-memory")
```

Comment:
This is maybe a bit frail at the moment.

The problem is that I can't apply the filter with lance because I am using a limit/offset and they won't be pushed down if a filter is present.

Potential solutions:
 * Add a scan_range to lance that interprets limit/offset as a pre-filter limit/offset.
 * Add filter support to take (and just pass a contiguous range to take)
 * Use duckdb or something else to apply the filter here